File: sdk/ruby/lib/shojiku/artifact.rb
# frozen_string_literal: true

module Shojiku
  # A rendered (and possibly signed) document.
  #
  # The application sees bytes and metadata — never a layout-engine internal,
  # and never a handle it has to free. Freeing is the binding's job and it is
  # already done by the time this object exists.
  class DocumentArtifact
    # The PDF, as binary. Always `ASCII-8BIT`: PDF bytes are not text and
    # tagging them as any character encoding is how a document gets corrupted
    # by a well-meaning transcode on the way to disk.
    attr_reader :bytes

    # How many pages the engine laid out. `nil` for an artifact that was
    # signed rather than rendered — signing appends a revision to bytes it
    # never measured, and a zero there would read as "a document with no
    # pages".
    attr_reader :page_count

    attr_reader :diagnostics

    # Where this document came from, which is what a strict client signs on:
    #
    # * `:rendered` — laid out from a template the configured root resolved,
    # * `:source` — laid out from template bytes the application supplied,
    # * `:loaded` — bytes the application supplied whole ({Client#artifact}).
    #
    # Only the first is signable under a {Lockdown}: in the other two the
    # provenance of what gets signed is the application's rather than the
    # deployment's, which is the distinction strict exists to draw. Signing
    # inherits the origin of what it signed — appending a revision does not
    # launder where the document came from. Verification is never restricted.
    attr_reader :origin

    # `origin` defaults to the LEAST privileged value, not the most: every
    # internal path states it explicitly, so the default only ever applies to
    # an artifact somebody built by hand — which is bytes handed over whole,
    # and must not become signable under a lockdown by omission.
    def initialize(bytes:, diagnostics:, client:, page_count: nil, origin: :loaded)
      @bytes = bytes
      @page_count = page_count
      @diagnostics = diagnostics
      @client = client
      @origin = origin
    end

    # Whether these bytes were handed over whole rather than laid out here.
    def loaded?
      @origin == :loaded
    end

    # Writes the document. Binary mode explicitly — a PDF contains NUL and
    # every other byte value, and text mode would translate line endings on
    # Windows.
    def write(path)
      File.binwrite(path, @bytes)
      path
    end

    def size
      @bytes.bytesize
    end

    # Signs this document, returning a {Result} carrying the signed artifact.
    # The signed bytes begin with these bytes byte for byte: signing appends a
    # revision, it never rewrites what was there.
    def sign(provider)
      @client.sign(self, provider)
    end

    # Verifies this document against caller-supplied trust anchors, returning
    # a {Result} carrying a {VerificationReport}.
    def verify(anchors: nil, anchors_pem: nil)
      @client.verify(self, anchors: anchors, anchors_pem: anchors_pem)
    end
  end
end

File: sdk/python/src/shojiku/artifact.py
"""A rendered (and possibly signed) document.

The application sees bytes and metadata — never a layout-engine internal, and
never a handle it has to free. Freeing is the binding's job and it is already
done by the time this object exists.
"""

from __future__ import annotations

from enum import StrEnum
from pathlib import Path
from typing import TYPE_CHECKING

from shojiku.diagnostic import Diagnostic

if TYPE_CHECKING:
    from shojiku.client import Client
    from shojiku.result import Result
    from shojiku.verification_report import VerificationReport


class Origin(StrEnum):
    """Where a document came from, which is what a strict client signs on.

    * ``RENDERED`` — laid out from a template the configured root resolved,
    * ``SOURCE`` — laid out from template bytes the application supplied,
    * ``LOADED`` — bytes the application supplied whole.

    Only the first is signable under a lockdown: in the other two the provenance
    of what gets signed is the application's rather than the deployment's, which
    is the distinction strict exists to draw. Signing inherits the origin of what
    it signed — appending a revision does not launder where the document came
    from. Verification is never restricted.
    """

    RENDERED = "rendered"
    SOURCE = "source"
    LOADED = "loaded"


class DocumentArtifact:
    """PDF bytes plus what the engine knows about them."""

    def __init__(
        self,
        bytes_: bytes,
        diagnostics: list[Diagnostic],
        client: Client,
        page_count: int | None = None,
        # The LEAST privileged value, not the most: every internal path states
        # it explicitly, so the default only ever applies to an artifact somebody
        # built by hand — which is bytes handed over whole, and must not become
        # signable under a lockdown by omission.
        origin: Origin = Origin.LOADED,
    ) -> None:
        # The PDF, as binary. PDF bytes are not text and decoding them to `str`
        # is how a document gets corrupted on the way to disk.
        self.bytes = bytes_
        # How many pages the engine laid out. None for an artifact that was
        # signed rather than rendered — signing appends a revision to bytes it
        # never measured, and a zero there would read as "a document with no
        # pages".
        self.page_count = page_count
        self.diagnostics = diagnostics
        self.origin = origin
        self._client = client

    @property
    def loaded(self) -> bool:
        """Whether these bytes were handed over whole rather than laid out here."""
        return self.origin == Origin.LOADED

    @property
    def size(self) -> int:
        return len(self.bytes)

    def write(self, path: str) -> str:
        """Write the document.

        Binary mode explicitly — a PDF contains NUL and every other byte value,
        and text mode would translate line endings on Windows.
        """
        Path(path).write_bytes(self.bytes)
        return path

    def sign(self, provider: object) -> Result[DocumentArtifact]:
        """Sign this document, returning a result carrying the signed artifact.

        The signed bytes begin with these bytes byte for byte: signing appends a
        revision, it never rewrites what was there.
        """
        return self._client.sign(self, provider)

    def verify(
        self,
        anchors: str | list[str] | None = None,
        anchors_pem: bytes | None = None,
    ) -> Result[VerificationReport]:
        """Verify this document against caller-supplied trust anchors."""
        return self._client.verify(self, anchors=anchors, anchors_pem=anchors_pem)
